File: 0048-rotate-image/0048-rotate-image.cpp
// TC- O(m*n)
// SC- O(1)

class Solution {
private:
    void transpose(vector<vector<int>>& matrix){
        for(int i=0; i<matrix.size(); i++){
            for(int j=i+1; j<matrix[0].size(); j++){
                swap(matrix[i][j], matrix[j][i]);
            }
        }
    }

    void reverse(vector<vector<int>>& matrix){
        for(int r=0; r<matrix.size(); r++){
            for(int i=0, j=matrix[0].size()-1; i<j; i++, j--){
                swap(matrix[r][i], matrix[r][j]);
            }
        }
    }
public:
    void rotate(vector<vector<int>>& matrix) {
        transpose(matrix);
        reverse(matrix);
    }
};

File: 0048-rotate-image/0048-rotate-image.py
class Solution:

    def transpose(self, matrix) -> None:
        for r in range(self.n):
            for c in range(r+1, self.n):
                if(r!=c):
                    matrix[r][c], matrix[c][r] = matrix[c][r], matrix[r][c]

    def reflect(self, matrix) -> None:
        for row in matrix:
            row.reverse()

    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        self.n = len(matrix)
        self.transpose(matrix)
        self.reflect(matrix)